### ISSUE TITLE
[1.13] docker version output is not consistent when there are downgrades or incompatibilities

### ISSUE THREAD

**Issue by febbraro** [docker/docker#30994]
When using `docker version` with the `--format` option to get specific pieces of data, there are cases when the data is not really parseable. If format is provided the command should respect the data that is requested and not litter the output with other information.

Example:
I have a docker-machine with docker version 1.11.2 (api version 1.23) and a docker client version 1.13 (api version 1.26)

```
>>> docker version --format {{.Client.APIVersion}}
1.24 (downgraded from 1.26)
Error response from daemon: client is newer than server (client API version: 1.24, server API version: 1.23)
```

In this scenario I expect 1.24 (or 1.26) as the response, and I expressly don't expect the error messages from the daemon included.

or another example

```
>>> docker version --format {{.Client.Version}}
1.13.1
Error response from daemon: client is newer than server (client API version: 1.24, server API version: 1.23)
```

In this scenario I expect 1.13.1 as the response, I was not requesting any info about the server.

I think providing the error and even the downgraded message are totally fine when I'm not requesting a specific piece of data with `--format`.  When building tools to help debug errors for my users this makes it very difficult (or impossible) to provide meaningful information about errors to the users when we cannot extract specific data because of those errors.

**Comment by thaJeztah** [comment]
> 

Agreed, the current implementation re-uses the "Version" type from the API (see https://github.com/docker/docker/blob/04f7a03359ac009ad2cd548fa93124a66e880ae5/cli/command/system/version.go#L78-L88). I think we should have a separate type that's used in the client, and have separate properties for "default" and "actual" API versions. [The template that's used for printing](https://github.com/docker/docker/blob/04f7a03359ac009ad2cd548fa93124a66e880ae5/cli/command/system/version.go#L18-L33) should then be modified accordingly


> and I expressly don't expect the error messages from the daemon included.

The error is expected when talking to an 1.11 daemon; API version negotiation was added in docker 1.13, and requires the daemon to return a header that's not present in docker 1.12 and older. For that reason, the client will not be able to automatically negotiate older versions, and will select API 1.24, which corresponds with the latest version of docker that does not send the header that's needed.

The error should be printed on `stderr`, so you can redirect those errors to suppress them (`docker version 2>/dev/null`)

I'll label this issue as "enhancement" and I welcome contributions that update the API-version output as I described above. I don't think the error printing has to be changed (unless it's not printed on `stderr`)

**Comment by thaJeztah** [comment]
FWIW, you can manually override the API version to older versions of the API using the `DOCKER_API_VERSION` environment variable. This variable is intended for debugging purposes though, as the client will not take older versions into account (i.e., it may attempt to use features that were not yet available in older daemons), for example;

```
DOCKER_API_VERSION=1.23 docker version

Client:
 Version:      1.14.0-dev
 API version:  1.23
 Go version:   go1.7.4
 Git commit:   7433634
 Built:        Mon Feb 13 22:56:54 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.14.0-dev
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.4
 Git commit:   7433634
 Built:        Mon Feb 13 22:56:54 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by jmzwcn** [comment]
#dibs

**Comment by jmzwcn** [comment]
Seems we could avoid un-expected info to just by adding
if err == nil {
		vd.Server = &serverVersion
	}
**else if(err include "client is newer than server"){**
fmt.Fprintf(os.Stderr,"client is newer than server: %v",err)
err=nil
}

@thaJeztah, any advice in detail?

**Comment by febbraro** [comment]
Thanks for the consideration, that would be great to have the data available. 

Also, you are correct, if I ignore stderr, I don't see the errors related to client and server and things work much more smoothly.

**Comment by thaJeztah** [comment]
@jmzwcn I think there's no special handling needed for the error; if it's an error, it's an error, so we should print it. But, fill me in if I misunderstood your comment :blush:. If you have an implementation, feel free to open as a pull request; we can always discuss there! 👍 

Thanks @febbraro, good to hear redirecting stderr works for that part; looks that @jmzwcn has volunteered to work on this (thanks!), so we'll be able to address the other bits in the 1.14 release. (I don't think we'd be able to include it in a patch release)

**Comment by jmzwcn** [comment]
Have create a [PR](https://github.com/docker/docker/pull/31022), please let me know if there is any problem.

**Comment by thaJeztah** [comment]
https://github.com/docker/docker/issues/31022 was merged, and will be in docker 17.05
